### PR TITLE
[libxml2] Update to latest commit, remove wrapper

### LINKED ISSUE #14468: libxml2 vcpkg-cmake-wrapper.cmake bugs
**Describe the bug**
In file vcpkg-cmake-wrapper.cmake of package LibXml2, the target LibXml2::LibXml2 is not linked against all necessary libraries

**Environment**
- OS: Windows

**To Reproduce**
1. ./vcpkg install libxml2
2. find_package(LibXml2)
target_link_libraries(main PRIVATE LibXml2::LibXml2)
3. ZLIB_LIBRARIES and LIBLZMA_LIBRARIES are not linked, neither ws2_32 on windows

**Expected behavior**
LibXml2::LibXml2 target should link with Iconv, Charset, lzma, zlib, ws2_32 (windows only) libraries

### PR DESCRIPTION
**Pull request by JackBoosY**
- Update to latest commit to use the internal `CMakeLists.txt`
- Remove `vcpkg-cmake-wrapper` and `usage` since the official already provide the cmake target

Replaced PR #14478.
Fixes #14468.

### CODE REVIEW

**Comment by traversaro** [comment]
FYI the upstream libxml2 project now has an official CMake build system (see https://gitlab.gnome.org/GNOME/libxml2/-/blob/master/CMakeLists.txt), and even the autotools build system installs a CMake config file (see https://gitlab.gnome.org/GNOME/libxml2/-/blob/master/libxml2-config.cmake.in). I am a bit afraid that adding a third vcpkg-specific variant will complicate life of libraries maintainers that need to support all this variants.

**Comment by JackBoosY** [comment]
@traversaro That's awesome, I will try to update libxml2 to use that.

**Comment by Neumann-A** [comment]
is there a reason not to switch to `vcpkg_configure_make` ? (Assuming there are problems with upstreams CMakeLists.txt)

**Comment by JackBoosY** [comment]
@Neumann-A If `CMakelists.txt` has no errors, we prefer to use cmake to build all ports.

File: ports/libxml2/fix-dependencies.patch [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+ set(LIBXML_VERSION_NUMBER ${LIBXML_VERSION})
+ 
+-find_package(Iconv)
+-find_package(ICU)
+-find_package(LibLZMA)
+-find_package(Python COMPONENTS Interpreter Development)
+-find_package(Threads)
+-find_package(ZLIB)
++if (LIBXML2_WITH_ICONV)
++find_package(unofficial-iconv CONFIG REQUIRED)
```
Comment (Neumann-A) [inline ports/libxml2/fix-dependencies.patch:16]:
```suggestion
+find_package(Iconv REQUIRED)
```

I don't like to fix it again in #11776. Use official cmake module here!

File: ports/libxml2/fix-dependencies.patch [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+-
+ set(LIBXML2_XMLCONF_WORKING_DIR ${CMAKE_CURRENT_BINARY_DIR} CACHE PATH "Working directory for XML Conformance Test Suite")
+ 
+ foreach(VARIABLE IN ITEMS WITH_AUTOMATA WITH_C14N WITH_CATALOG WITH_DEBUG WITH_DOCB WITH_EXPR WITH_FTP WITH_HTML WITH_HTTP WITH_ICONV WITH_ICU WITH_ISO8859X WITH_LEGACY WITH_LZMA WITH_MEM_DEBUG WITH_MODULES WITH_OUTPUT WITH_PATTERN WITH_PUSH WITH_READER WITH_REGEXPS WITH_RUN_DEBUG WITH_SAX1 WITH_SCHEMAS WITH_SCHEMATRON WITH_THREADS WITH_THREAD_ALLOC WITH_TREE WITH_TRIO WITH_UNICODE WITH_VALID WITH_WRITER WITH_XINCLUDE WITH_XPATH WITH_XPTR WITH_ZLIB)
+@@ -456,7 +438,7 @@ if(WIN32)
+ endif()
+ 
+ if(LIBXML2_WITH_ICONV)
+-	target_link_libraries(LibXml2 PUBLIC Iconv::Iconv)
++	target_link_libraries(LibXml2 PUBLIC unofficial::iconv::libiconv)
```
Comment (Neumann-A) [inline ports/libxml2/fix-dependencies.patch:95]:
```suggestion
+	target_link_libraries(LibXml2 PUBLIC Iconv::Iconv)
```

File: ports/libxml2/fix-dependencies.patch [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
++++ b/libxml2-config.cmake.cmake.in
+@@ -56,33 +56,30 @@ select_library_location(LibXml2::LibXml2 LIBXML2)
+ include(CMakeFindDependencyMacro)
+ 
+ if(@LIBXML2_WITH_ICONV@)
+-  find_dependency(Iconv)
+-  list(APPEND LIBXML2_LIBRARIES    ${Iconv_LIBRARIES})
+-  list(APPEND LIBXML2_INCLUDE_DIRS ${Iconv_INCLUDE_DIRS})
++  find_dependency(unofficial-iconv CONFIG REQUIRED)
++  list(APPEND LIBXML2_LIBRARIES    unofficial::iconv::libiconv)
```
Comment (Neumann-A) [inline ports/libxml2/fix-dependencies.patch:126]:
```suggestion
+  find_dependency(Iconv REQUIRED)
+  list(APPEND LIBXML2_LIBRARIES    Iconv::Iconv)
```

File: ports/libxml2/fix-dependencies.patch [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+ endif()
+ 
+ if(NOT @BUILD_SHARED_LIBS@)
+   if(@LIBXML2_WITH_THREADS@)
+-    find_dependency(Threads)
++    find_dependency(Threads REQUIRED)
```
Comment (Neumann-A) [inline ports/libxml2/fix-dependencies.patch:132]: >  find_dependency(Threads REQUIRED)

please reread find_dependency docs. You are not supposed to pass `REQUIRED`

Comment (JackBoosY) [reply ports/libxml2/fix-dependencies.patch:132]: `find_dependency forwards the correct parameters for QUIET and REQUIRED which were passed to the original find_package() call`

File: ports/libxml2/fix-dependencies.patch (hunk 1 above)
Comment (JackBoosY) [reply ports/libxml2/fix-dependencies.patch]: `x64-windows-static` build requires `iconv::charset`, but `FindIconv.cmake` does not provide it. So we cannot use module mode.

Comment (Neumann-A) [reply ports/libxml2/fix-dependencies.patch:16]: https://github.com/JackBoosY/vcpkg/pull/1 ..... just fix the damn wrapper. No `iconv::charset` is required just linking against charset.lib in the `Iconv::Iconv` target....
There are times where targets make sense and there are times where just fixing the linkage is enough. Targets are a cmake concept while linkage is a linker concept and only the latter is really required for a succesful build and consumption in downstream ports.

File: ports/libspatialite/portfile.cmake [hunk 5]
```diff
@@ -15,25 +15,27 @@ if (VCPKG_TARGET_IS_WINDOWS)
...
     else()
         set(CL_FLAGS_DBG "/MTd /Zi /DACCEPT_USE_OF_DEPRECATED_PROJ_API_H")
         set(CL_FLAGS_REL "/MT /Ox /DACCEPT_USE_OF_DEPRECATED_PROJ_API_H")
         set(GEOS_LIBS_REL "${CURRENT_INSTALLED_DIR}/lib/geos_c.lib ${CURRENT_INSTALLED_DIR}/lib/geos.lib")
         set(GEOS_LIBS_DBG "${CURRENT_INSTALLED_DIR}/debug/lib/geos_cd.lib ${CURRENT_INSTALLED_DIR}/debug/lib/geosd.lib")
-        set(LIBXML2_LIBS_REL "${CURRENT_INSTALLED_DIR}/lib/libxml2.lib ${CURRENT_INSTALLED_DIR}/lib/lzma.lib ws2_32.lib")
-        set(LIBXML2_LIBS_DBG "${CURRENT_INSTALLED_DIR}/debug/lib/libxml2.lib ${CURRENT_INSTALLED_DIR}/debug/lib/lzmad.lib ws2_32.lib")
+        set(LIBXML2_LIBS_REL "${CURRENT_INSTALLED_DIR}/lib/libxml2s.lib ${CURRENT_INSTALLED_DIR}/lib/lzma.lib ws2_32.lib")
+        set(LIBXML2_LIBS_DBG "${CURRENT_INSTALLED_DIR}/debug/lib/libxml2ds.lib ${CURRENT_INSTALLED_DIR}/debug/lib/lzmad.lib ws2_32.lib")
     endif()
```
Comment (Neumann-A) [inline ports/libspatialite/portfile.cmake:39]: The whole section looks incorrect and is mixing `VCPKG_CRT_LINKAGE` and `VCPKG_LIBRARY_LINKAGE`. Use find_library to find the libraries.

**Comment by Neumann-A** [comment]
no no no..... don't do that. You are patching the hell out of downstream ports for some unnecessary reason. Fix the CMake `FindLibXml2.cmake` module with a `vcpkg-cmake-wrapper.cmake` and be done with it!

You can override all manual `find_library` calls by doing a `find_library` call yourself before the `find_package()` call since `find_library` sets a CACHE variable.

**Comment by JackBoosY** [comment]
@Neumann-A I've already realized that...
The official provided cmake is incomplete, I think we should wait for upstream fix these issues.

**Pull request closed without merging**